apps/prairielearn/src/migrations/20250707213723_audit_events_add.sql

```diff
@@ -0,0 +1,45 @@
+CREATE TYPE audit_event_action AS ENUM ('insert', 'update', 'delete');
+
+CREATE TABLE IF NOT EXISTS audit_events (
+  id BIGSERIAL PRIMARY KEY,
+  date TIMESTAMP WITH TIME ZONE DEFAULT CURRENT_TIMESTAMP,
+  action audit_event_action NOT NULL,
+  action_detail TEXT,
+  table_name TEXT NOT NULL,
+  agent_authn_user_id BIGINT,
+  agent_user_id BIGINT,
+  institution_id BIGINT,
+  course_id BIGINT,
+  course_instance_id BIGINT,
+  assessment_id BIGINT,
+  assessment_instance_id BIGINT,
+  assessment_question_id BIGINT,
+  group_id BIGINT,
+  subject_user_id BIGINT,
+  context JSONB NOT NULL DEFAULT '{}'::jsonb,
+  old_row JSONB,
+  new_row JSONB,
+  row_id BIGINT NOT NULL,
+  CONSTRAINT audit_events_agent_authn_user_id_fkey FOREIGN KEY (agent_authn_user_id) REFERENCES users (user_id) ON UPDATE CASCADE ON DELETE SET NULL,
+  CONSTRAINT audit_events_agent_user_id_fkey FOREIGN KEY (agent_user_id) REFERENCES users (user_id) ON UPDATE CASCADE ON DELETE SET NULL,
+  CONSTRAINT audit_events_course_id_fkey FOREIGN KEY (course_id) REFERENCES pl_courses (id) ON UPDATE CASCADE ON DELETE SET NULL,
+  CONSTRAINT audit_events_course_instance_id_fkey FOREIGN KEY (course_instance_id) REFERENCES course_instances (id) ON UPDATE CASCADE ON DELETE SET NULL,
+  CONSTRAINT audit_events_group_id_fkey FOREIGN KEY (group_id) REFERENCES groups (id) ON UPDATE CASCADE ON DELETE SET NULL,
+  CONSTRAINT audit_events_subject_user_id_fkey FOREIGN KEY (subject_user_id) REFERENCES users (user_id) ON UPDATE CASCADE ON DELETE SET NULL,
+  CONSTRAINT audit_events_institution_id_fkey FOREIGN KEY (institution_id) REFERENCES institutions (id) ON UPDATE CASCADE ON DELETE SET NULL,
+  CONSTRAINT audit_events_assessment_id_fkey FOREIGN KEY (assessment_id) REFERENCES assessments (id) ON UPDATE CASCADE ON DELETE SET NULL,
```

Comment:
The more I look at this, I'm getting nervous about these being actual foreign keys. If we delete groups, or assessment instances, or assessment questions, we'll end up doing a table scan to find rows to delete, which will get very slow as this table gets big.

I think we either need indexes on any tables for which we do hard-deletes, or we need to make these not foreign keys like we do in PrairieTest.

Reply:
> Especially for audit logs, there are benefits to making them immutable and retaining them even when someone deletes a record that they're associated with. This can be really helpful at times for forensics.

If this is the argument for retaining them, can't we just drop the referential integrity on the two columns that can be hard deleted?